### PR TITLE
Jwakefield/add role management

### PR DESCRIPTION
**Pull request by JoelWakefield**
- view roles
- view users
- create roles
- add roles to user
- remove roles from user

### CODE REVIEW

File: ProjectManagementApp/Program.cs [hunk 1]
```diff
@@ -29,12 +30,13 @@
...
 
-builder.Services.AddIdentityCore<ApplicationUser>(options => options.SignIn.RequireConfirmedAccount = true)
+builder.Services.AddIdentityCore<ApplicationUser>(options => options.SignIn.RequireConfirmedAccount = false)
+    .AddRoles<IdentityRole>()
     .AddEntityFrameworkStores<ApplicationDbContext>()
     .AddSignInManager()
     .AddDefaultTokenProviders();
 
-builder.Services.AddSingleton<IEmailSender<ApplicationUser>, IdentityNoOpEmailSender>();
+//builder.Services.AddSingleton<IEmailSender<ApplicationUser>, IdentityNoOpEmailSender>();
```
Comment (jwthurmond) [inline ProjectManagementApp/Program.cs:39]: remove commented code

File: ProjectManagementApp/Program.cs [hunk 2]
```diff
@@ -52,6 +54,14 @@
     app.UseHsts();
 }
 
+using var serviceScope = app.Services.GetService<IServiceScopeFactory>()!.CreateScope();
+//var userManager = serviceScope.ServiceProvider.GetRequiredService<UserManager<ApplicationUser>>();
```
Comment (jwthurmond) [inline ProjectManagementApp/Program.cs:58]: remove commented code

File: README.md [hunk 3]
```diff
@@ -21,4 +21,14 @@ dotnet ef migrations add <migration name>
 - Update the database
 ```
 dotnet ef database update 
+```
+
+## Adding Roles
+> NOTE: When using `AddIdentityCore`, it's important to call the `AddRoles` before `AddEntityFrameworkStores`.
```
Comment (jwthurmond) [inline README.md:27]: interesting, nice find

**Review by jwthurmond — APPROVED** [review]
Looks fine, I'd remove the commented code unless it's something you're going to use later on.

File: ProjectManagementApp/Program.cs (hunk 2 above)
Comment (JoelWakefield) [reply ProjectManagementApp/Program.cs]: I've put all the prebuilt roles and users into the `UserBuilderService`; this should build only on the first time the app is run.

File: ProjectManagementApp/Program.cs (hunk 1 above)
Comment (JoelWakefield) [reply ProjectManagementApp/Program.cs]: it's gone - I think I'll keep some of the default identity services for now (I might use them later, but they're harmless to leave in).

File: ProjectManagementApp/Services/UserBuilderService.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            if (roleManager.RoleExistsAsync(Roles.Admin).Result == false)
+                roleManager.CreateAsync(new IdentityRole(Roles.Admin)).Wait();
+            if (roleManager.RoleExistsAsync(Roles.Manager).Result == false)
+                roleManager.CreateAsync(new IdentityRole(Roles.Manager)).Wait();
+            if (roleManager.RoleExistsAsync(Roles.Lead).Result == false)
+                roleManager.CreateAsync(new IdentityRole(Roles.Lead)).Wait();
+            if (roleManager.RoleExistsAsync(Roles.Member).Result == false)
+                roleManager.CreateAsync(new IdentityRole(Roles.Member)).Wait();
+
+            // Create initial roles
```
Comment (jwthurmond) [inline ProjectManagementApp/Services/UserBuilderService.cs:24]: // Create initial roles -> //Create initial users

**Review by jwthurmond — APPROVED** [review]
just a small change to the comment

File: ProjectManagementApp/Services/UserBuilderService.cs (hunk 4 above)
Comment (JoelWakefield) [reply ProjectManagementApp/Services/UserBuilderService.cs]: fixed

**Pull request merged**